File: Sihyeon/Week4/1915_가장큰정사각형.py
n, m = map(int, input().split())
arr = []

# 정사각형 입력받기
for i in range(n):
    row = list(map(int, input()))
    arr.append(row)

answer = 0

# 배열이 1x1인 경우
if n == 1 and m == 1:
    answer = arr[0][0]
else:
    # arr 요소 중 값이 1일 때 arr의 전행, 전열, 전행전열 비교해서 가장 작은 것에서 1을 더한다.
    for i in range(1, n):
        for j in range(1, m):
            if arr[i][j] == 1:
                arr[i][j] = min(arr[i-1][j], arr[i][j-1], arr[i-1][j-1]) + 1
                answer = max(answer, arr[i][j])

print(answer * answer)

File: Jihun/week_4/1915_가장큰정사각형.cpp
#include <iostream>
using namespace std;

int min2(int a, int b, int c) { return min(a, min(b, c)); }

int main() {
    int arr[1001][1001];
    int dp[1001][1001];
    int n, m;
    cin >> n >> m;
    int result = 0;

    for (int i = 0; i < n; i++) {
        for (int j = 0; j < m; j++) {
            char c;
            cin >> c;
            arr[i][j] = c - '0';
        }
    }
    for (int i = 0; i < n; i++) {
        dp[i][0] = arr[i][0];
        result = max(arr[i][0], result);
    }

    for (int i = 0; i < m; i++) {
        dp[0][i] = arr[0][i];
        result = max(arr[0][i], result);
    }

    for (int i = 1; i < n; i++) {
        for (int j = 1; j < m; j++) {
            if (arr[i][j] == 1)
                dp[i][j] =
                    min2(dp[i][j - 1], dp[i - 1][j - 1], dp[i - 1][j]) + 1;
            result = max(dp[i][j], result);
        }
    }

    cout << result * result << endl;

    return 0;
}